### PR TITLE
Archive resource history - part I

### PR DESCRIPTION
**Pull request by anatbal**
# Resolves part of #2491 

This is the first part of the implementation for archiving a resource history

Creation of a new container name resources_history
Creation of ResourceHistoryRepository class
Save each history item in the new container (maintain the existing behavior, this will be removed in the next PR together with migrations)
Note - These changes are not touching airlock just yet

### CODE REVIEW

File: api_app/db/repositories/resources_history.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+class ResourceHistoryRepository(BaseRepository):
+    @classmethod
+    async def create(cls, client: CosmosClient):
+        cls = ResourceHistoryRepository()
+        await super().create(client, config.STATE_STORE_RESOURCES_HISTORY_CONTAINER)
+        return cls
+
+    @staticmethod
+    def resource_history_query(resourceId: str):
+        return f'SELECT * FROM c WHERE c.resourceId = "{resourceId}"'
```
Comment (yuvalyaron) [inline api_app/db/repositories/resources_history.py:19]: do we need this method?
since we know the ID, it's better to not use SQL, and just read it directly -
it's cheaper in terms of RU.

**Review by yuvalyaron — CHANGES_REQUESTED** [review]

File: api_app/models/domain/resource.py [hunk 2]
```diff
@@ -16,14 +16,28 @@ class ResourceType(str, Enum):
     SharedService = strings.RESOURCE_TYPE_SHARED_SERVICE
 
 
+class NewResourceHistoryItem(AzureTREModel):
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/models/domain/resource.py:19]: No. This is strictly for making it easier for review

File: api_app/api/routes/airlock_resource_helpers.py [hunk 3]
```diff
@@ -147,6 +148,7 @@ async def delete_review_user_resource(
         workspace_service_repo: WorkspaceServiceRepository,
         resource_template_repo: ResourceTemplateRepository,
         operations_repo: OperationRepository,
+        resource_history_repo: ResourceHistoryRepository,
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/api/routes/airlock_resource_helpers.py:151]: I strongly disagree.  Since the UI is using the history this PR will become huge and no way anyone can review it. Had this discussion with @tamirkamara  as well that this is going to be divided.

File: api_app/db/repositories/resources.py [hunk 4]
```diff
@@ -112,7 +113,8 @@ async def patch_resource(self, resource: Resource, resource_patch: ResourcePatch
             templateVersion=resource_copy.templateVersion
         )
         resource.history.append(history_item)
-
+        # create the history item in the resource history repository
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/db/repositories/resources.py:116]: Again, only for reviewer to be able to see the difference

File: api_app/models/domain/resource.py [hunk 5]
```diff
@@ -16,14 +16,28 @@ class ResourceType(str, Enum):
...
+    """
+    Resource History Item - to preserve history of resource properties
+    """
+    id: str = Field(title="Id", description="GUID identifying the resource request")
+    resourceId: str = Field(title="Id", description="GUID identifying the resource request")
```
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/models/domain/resource.py:24]: Typo will fix

File: api_app/db/repositories/resources_history.py (hunk 1 above)
Comment (anatbal) [reply api_app/db/repositories/resources_history.py:19]: I understand what you are saying, but here it's different. we are doing the query by **resourceId** which means it will potentially bring a list of items, since these are all the history items that are related to this resource, each will have a different **id**. The thing you commented on is not the function but strictly a  const for the sql query which is necessary.

File: api_app/models/domain/resource.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eladiw) [reply api_app/models/domain/resource.py:19]: so, what is the name? ResourceHistoryItem?
I suggest you already rename

File: api_app/models/domain/resource.py [hunk 6]
```diff
@@ -16,14 +16,28 @@ class ResourceType(str, Enum):
...
 class ResourceHistoryItem(AzureTREModel):
     """
     Resource History Item - to preserve history of resource properties
     """
-    properties: dict = {}
+    properties: dict = Field({}, title="Resource template parameters", description="Parameters for the deployment")
     isEnabled: bool
     resourceVersion: int
-    updatedWhen: float
+    updatedWhen: float = 0
```
Comment (eladiw) [inline api_app/models/domain/resource.py:40]: why is this needed

File: api_app/core/config.py [hunk 7]
```diff
@@ -26,6 +26,7 @@
 STATE_STORE_DATABASE = "AzureTRE"
 STATE_STORE_RESOURCES_CONTAINER = "Resources"
 STATE_STORE_RESOURCE_TEMPLATES_CONTAINER = "ResourceTemplates"
+STATE_STORE_RESOURCES_HISTORY_CONTAINER = "ResourcesHistory"
```
Comment (eladiw) [inline api_app/core/config.py:29]: up to debate.
ResourceTemplates --> implies --> ResourceHistories?

File: api_app/db/repositories/resources_history.py (hunk 1 above)
Comment (eladiw) [reply api_app/db/repositories/resources_history.py:19]: please make sure. at least somewhere along the process, this resource id is sanitized.

File: api_app/db/repositories/resources_history.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    @staticmethod
+    def resource_history_with_resource_version_query(resourceId: str, resourceVersion: str):
+        return f'SELECT * FROM c WHERE c.resourceId = "{resourceId}" AND c.resourceVersion = "{resourceVersion}"'
```
Comment (eladiw) [inline api_app/db/repositories/resources_history.py:23]: same comment

File: api_app/db/repositories/resources_history.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    async def get_resource_history_by_resource_id(self, resourceId: str) -> List[NewResourceHistoryItem]:
+        query = self.resource_history_query(resourceId)
+        resource_history_items = await self.query(query=query)
```
Comment (eladiw) [inline api_app/db/repositories/resources_history.py:27]: can we add some logs? e.g. number of items in the list. or more useful debug info?

File: api_app/db/repositories/resources_history.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        return parse_obj_as(List[NewResourceHistoryItem], resource_history_items)
```
Comment (eladiw) [inline api_app/db/repositories/resources_history.py:28]: what happens if this parse fails?

**Review by eladiw — CHANGES_REQUESTED** [review]
Looks good. some comments.
in general can you add more logs?

File: api_app/api/routes/airlock_resource_helpers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anatbal) [reply api_app/api/routes/airlock_resource_helpers.py:151]: Since the UI is using the history this PR will become huge and no way anyone can review it. Had this discussion with @tamirkamara as well that this is going to be divided.
Regarding the usage, it has all the details in the PR description

File: api_app/db/repositories/resources_history.py (hunk 1 above)
Comment (yuvalyaron) [reply api_app/db/repositories/resources_history.py:19]: oh, I see, but if you query based on the resource ID, and the partition key is the ID -  shouldn't we choose the resource ID to be the partition key? this way the history items for a resource will be stored on the same partition, which will be better in terms of cost and performance

File: api_app/db/repositories/resources.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuvalyaron) [reply api_app/db/repositories/resources.py:116]: your call, but I don't find it necessary

File: api_app/db/repositories/resources_history.py (hunk 1 above)
Comment (anatbal) [reply api_app/db/repositories/resources_history.py:19]: I agree. changing the partition key to resourceID.
Just documenting here that even though the operation repo acts in a similar manner it has a partition key of id, most likely because no one noticed.

File: api_app/models/domain/resource.py (hunk 6 above)
Comment (anatbal) [reply api_app/models/domain/resource.py:40]: To match it with default values like the rest of the models

File: api_app/core/config.py (hunk 7 above)
Comment (anatbal) [reply api_app/core/config.py:29]: after a discussion ResourceHistory was picked

File: api_app/db/repositories/resources_history.py (hunk 8 above)
Comment (anatbal) [reply api_app/db/repositories/resources_history.py:23]: Will add validation for special chars and guid

File: api_app/db/repositories/resources_history.py (hunk 10 above)
Comment (anatbal) [reply api_app/db/repositories/resources_history.py:28]: An error will be thrown similar to all our routes

**Comment by anatbal** [comment]
I am closing this since I opened a new PR. All the comments are fixed there
